src/Storages/MergeTree/MergeTreeData.cpp

```diff
@@ -5549,12 +5549,16 @@ class MergeTreeData::RestoredPartsHolder
         attachIfAllPartsRestored();
     }
 
-    String getTemporaryDirectory(const DiskPtr & disk)
+    String getTemporaryDirectory(const DiskPtr & disk, const String & part_name)
     {
         std::lock_guard lock{mutex};
-        auto it = temp_dirs.find(disk);
-        if (it == temp_dirs.end())
-            it = temp_dirs.emplace(disk, std::make_shared<TemporaryFileOnDisk>(disk, "tmp/")).first;
+        auto it = temp_part_dirs.find(part_name);
+        if (it == temp_part_dirs.end())
+        {
+            it = temp_part_dirs.emplace(part_name, std::make_shared<TemporaryFileOnDisk>(
+                disk,
+                fs::path{storage->getFullPathOnDisk(disk)} / ("tmp_restore_" + part_name + "-"))).first;
```

Comment:
This is the second part of the fix. During RESTORE it copies parts from a backup to a temporary directories on the destination disks, and then attaches those parts to tables.

Earlies it created those temporary directories as subfolders of the `/tmp/` folder on destination disks. That didn't work good for `s3_plain_rewritable` because that kind of disk can't move a folder to any place, it can only rename it (keeping its parent the same).

So I changed that in this PR and now temporary directories are created in the data folder of a destination table, for example
```
/var/lib/clickhouse/data/db_name/table_name/tmp_restore_all_0_1_1_0-XXXXXXXX
```
So to attach such folders they have to be renamed only (and not to be moved to another location).

Reply:
I'm somewhat worried this change affects all disk types, not just plain_rewritable, though I can't think of any actual pitfalls. The advantage of the `/tmp/` folder is that it provides a clear separation between temporary and persistent fs paths. Maybe scope this change to the `plain_rewritable` disk only?

Reply:
Actually the new way of naming temporary directories corresponds to how temporary directories have always been named by the `INSERT` command:
```
/var/lib/clickhouse/data/db_name/table_name/tmp_insert_all_0_1_1_0
```